javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/HelperReferenceWrapper.java

```diff
@@ -0,0 +1,278 @@
+    private final Map<String, Reference> helperReferences;
+
+    public Factory(TypePool classpathPool, Map<String, Reference> helperReferences) {
+      this.classpathPool = classpathPool;
+      this.helperReferences = helperReferences;
+    }
+
+    public HelperReferenceWrapper create(Reference reference) {
+      return new ReferenceType(reference);
+    }
+
+    private HelperReferenceWrapper create(String className) {
+      if (helperReferences.containsKey(className)) {
+        return new ReferenceType(helperReferences.get(className));
+      }
+      Resolution resolution = classpathPool.describe(className);
+      if (resolution.isResolved()) {
+        return new ClasspathType(resolution.resolve());
+      }
+      throw new IllegalStateException("Missing class " + className);
+    }
+
+    private final class ReferenceType implements HelperReferenceWrapper {
+      private final Reference reference;
+
+      private ReferenceType(Reference reference) {
+        this.reference = reference;
+      }
+
+      @Override
+      public boolean isAbstract() {
+        return reference.getFlags().contains(Flag.ABSTRACT);
+      }
+
+      @Override
+      public boolean hasSuperTypes() {
+        return hasActualSuperType() || reference.getInterfaces().size() > 0;
+      }
+
+      // Uses guava iterables to avoid unnecessary collection copying
+      @Override
+      public Iterable<HelperReferenceWrapper> getSuperTypes() {
+        Iterable<HelperReferenceWrapper> superClass = emptyList();
+        if (hasActualSuperType()) {
+          superClass = singleton(Factory.this.create(reference.getSuperName()));
+        }
+
+        Iterable<HelperReferenceWrapper> interfaces =
+            FluentIterable.from(reference.getInterfaces()).transform(toWrapper());
+
+        return Iterables.concat(superClass, interfaces);
+      }
+
+      private boolean hasActualSuperType() {
+        return reference.getSuperName() != null
+            && !reference.getSuperName().equals(Object.class.getName());
+      }
+
+      private Function<String, HelperReferenceWrapper> toWrapper() {
+        return new Function<String, HelperReferenceWrapper>() {
+          @Override
+          public HelperReferenceWrapper apply(String interfaceName) {
+            return Factory.this.create(interfaceName);
+          }
+        };
+      }
+
+      // Uses guava iterables to avoid unnecessary collection copying
+      @Override
+      public Iterable<Method> getMethods() {
+        return FluentIterable.from(reference.getMethods())
+            .filter(isOverrideable())
+            .transform(toMethod());
+      }
+
+      private Predicate<Reference.Method> isOverrideable() {
+        return new Predicate<Reference.Method>() {
+          @Override
+          public boolean apply(Reference.Method input) {
+            return !(input.getFlags().contains(Flag.STATIC)
+                || input.getFlags().contains(Flag.PRIVATE_OR_HIGHER)
```

Comment:
Is private method overridable?

Reply:
Nope, they're not - so we can skip checking them.

Reply:
Now I am even more confused. We create a predicate for `isOverrideable` and we say it is `true` when method 

- is not static AND
- is not private or higher AND
- is not a constructor

aren't ALL methods "private or higher"? Which makes this predicate constantly `false`?

Reply:
Ah, that's what you meant. The problem here is that `PRIVATE_OR_HIGHER` (and others) is used in two ways in muzzle:
1) it just simply marks that a method/class is private which is the case here, or
2) it describes a minimum access level required between a type and a method/field/class.

I'll split those two types of usage into separate flags to make it a bit more clear.